torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,906 @@
+
+    def get_tree_manager(self) -> CUDAGraphTreeManager:
+        with self.lock:
+            if self.tree_manager is None:
+                self.tree_manager = CUDAGraphTreeManager()
+            return self.tree_manager
+
+
+local = threading.local()
+local.tree_manager_container = TreeManagerContainer()
+
+# We need to register this as an object that will be copied over as TLS when new
+# threads are created in autograd
+torch._C._stash_obj_in_tls("tree_manager_container", local.tree_manager_container)
+
+
+def get_container():
+    if hasattr(local, "tree_manager_container"):
+        return local.tree_manager_container
+    assert torch._C._is_key_in_tls("tree_manager_container")
+    return torch._C._get_obj_in_tls("tree_manager_container")
+
+
+def cudagraphify_impl(model, inputs, static_input_idxs=()):
+    manager = get_container().get_tree_manager()
+    return manager.add_function(model, inputs, static_input_idxs)
+
+
+def is_live(weak_ref):
+    if weak_ref is None:
+        return False
+    return weak_ref() is not None
+
+
+class TensorWeakRef:
+    """
+    Wrapper around a weak ref of a Tensor that handles the _fix_weakref() call required
+    when unwrapping a Tensor weakref.
+    """
+
+    ref: WeakRef[Tensor]
+
+    def __init__(self, tensor: Tensor):
+        assert isinstance(tensor, Tensor)
+        self.ref = weakref.ref(tensor)
+
+    def __call__(self):
+        out = self.ref()
+        if out is None:
+            return out
+        assert isinstance(out, Tensor)
+        out._fix_weakref()
+        return out
+
+
+def is_cuda_tensor(x):
+    return isinstance(x, torch.Tensor) and x.device.type == "cuda"
+
+
+class CUDAGraphNode:
+    """
+    A single recording of a function into a CUDA Graph. Recordings of CUDA Graphs share a single memory pool
+    and are structured into a tree, where there is a single recording that can precede it (parent) and multiple
+    subsequent recordings that may follow (children). A node will have no parent if it is the first recording
+    in a tree; i.e., when it is first recorded, there are no live tensors from a previous recording which
+    would force a dependency.
+
+    On first recording, all of the live tensors in the current CUDA Graph Node path will be
+    reflected in the corresponding private pool. On subsequent executions, the caching allocator
+    is unaffected when the graph is replayed.
+
+    In order to support recording a subsequent cuda graph recording after execution of this graph,
+    we checkpoint the state of the memory pool so that it may later be resumed.
+
+    See [setCheckpointPoolState] for further explanation, as well as
+    https://user-images.githubusercontent.com/13564/222815509-374f3400-f83d-4f7d-8fa6-4a092b3250bb.png
+    """
+
+    def __init__(
+        self,
+        wrapped_function: WrappedFunction,
+        id: GraphID,
+        parent: Optional[CUDAGraphNode],
+        inputs: List[Tensor],
+        cuda_graphs_pool: Tuple[int, int],
```

Comment:
What's this

Reply:
lol I was surprised by that too, it's an `Opaque token (returned by a call to [graph_pool_handle()] or [other_Graph_instance.pool()])` according to the doc, but it looks like an `(int, int)`

Reply:
This is the memory pool handle to use a single pool across separate recordings